docs/design_proposals/port-forward.md

```diff
@@ -0,0 +1,145 @@
+has a [service on port 8080](https://github.com/GoogleContainerTools/skaffold/tree/HEAD/examples/react-reload/k8s/deployment.yaml#L7)
+and a [deployment/pod on port 8080](https://github.com/GoogleContainerTools/skaffold/blob/master/examples/react-reload/k8s/deployment.yaml#L29).
+
+
+
+## Design
+
+This document proposes to change Skaffold port-forwarding's defaults in a similar fashion as proposed
+by @corneliusweig in https://github.com/GoogleContainerTools/skaffold/issues/1564#issuecomment-473528574.
+These changes are intended to be backwards-compatible with the previous version
+of Skaffold except where noted.
+
+
+### Command-Line Changes
+
+> **NOTE**: Ideally we would re-purpose Skaffold's existing `--port-forward` CLI argument in
+> a backward-compatible manner (still under investigation).  This document assumes it is
+> not possible.
+
+~Skaffold's `--port-forward` argument should be changed from a binary true/false option to~
+
+Skaffold will introduce a new `--port-forward-modes` argument that takes
+a set of comma-separated values with the following defined modes:
+
+   - `user`: user-defined port-forwards as defined in the `skaffold.yaml`
+   - `services`: service ports are forwarded
+   - `pods`: `containerPort`s defined on Pods and Kubernetes workload objects that have pod-specs
+     are forwarded (Deployment, ReplicaSet, StatefulSet, DaemonSet, Job, CronJob)
+   - `debug`: an internal strategy to forward debugging-related ports on Pods as set up
+     from `skaffold debug`
+   - `off`: no ports are ever forwarded
+
+#### Changes to port-forwarding defaults
+
+Skaffold will default to enabling port-forwarding for establishing user-defined port-forwards
+as defined in the `skaffold.yaml`.  Skaffold's `dev` and `debug` will be changed to the following defaults.
+
+Command-line                        | v1.15.0          | New default modes
+----------------------------------- | ---------------- | -------------------
+`skaffold dev`                        | off            | user
+`skaffold dev --port-forward`         | user, services | user, services (no change)
+`skaffold dev --port-forward=false`   | off            | off (no change)
+`skaffold debug`                      | off            | user, debug
+`skaffold debug --port-forward`       | user, services, pods | user, services, debug
+`skaffold debug --port-forward=false` | off            | off (no change)
+
+**Compatibility Change:** The behaviour of `skaffold debug --port-forward` no longer forwards all
+`containerPort`s defined on pods.  Container ports were forwarded as there was no other option to forward
+just debug ports.  With this proposed change, the `debug` mode will only select debug ports.  Forwarding all
+`containerPort`s was the cause of some confusion in earlier iterations of port-forwarding.
+
+
+### Skaffold configuration changes
+
+#### skaffold.yaml
+
+We will define a new configuration block `defaults` with a single array field `portForwardModes`:
```

Comment:
is this meant to be extended in the future? seems strange to add a top-level `defaults` field if there's only port-related config in it. maybe we can add this in the `portForward` stanza instead?

Reply:
I had been thinking that we could put other defaults in here, like `default-repo` (#2317), but most of our other settings are found under the appropriate area.

The `portForward` block is an array of definitions, and I didn't think of an obvious field name to tuck them under.  `definitions`?  `resources`?
```yaml
portForward:
  defaultModes: ["..."]
  resources:
    - resourceType: service
       resourceName: example
```